File: src/native/NativeLabModule.swift
import ExpoModulesCore
import UIKit

public class NativeLabModule: Module {
  public func definition() -> ModuleDefinition {
    // JavaScript can call this synchronously.
    Function("greet") { (name: String) -> String in
      return "Hello \(name)! This message came from Swift."
    }

    // UIKit-related work should run on the main queue.
    AsyncFunction("getDeviceInfo") { () -> [String: String] in
      return [
        "name": UIDevice.current.name,
        "system": UIDevice.current.systemName,
        "version": UIDevice.current.systemVersion
      ]
    }
    .runOnQueue(.main)
  }
} 

File: src/native/NativeLabModule.ts
import { requireNativeModule } from "expo-modules-core";

export type DeviceInfo = {
  name: string;
  system: string;
  version: string;
};

type NativeLabModuleType = {
  greet(name: string): string;
  getDeviceInfo(): Promise<DeviceInfo>;
};

export default requireNativeModule("NativeLabModule") as NativeLabModuleType;
